File: apps/web/lib/process-job.ts
/** Poll async papercraft processing jobs until complete or failed. */

import type { CraftabilityScore, ProcessStats, ProjectStatus } from "@/types";

import { apiAuthHeaders } from "@/lib/api-auth";
import { streamJobEvents, withStreamOrFallback } from "@/lib/job-stream";
import { abortableDelay, throwIfAborted } from "@/lib/poll-utils";
import { ProcessJobCancelledError } from "@/lib/process-errors";

export type ProcessJobResponse = {
  jobId: string;
  projectId: string;
  status: "queued" | "running" | "completed" | "failed" | "cancelled";
  progress: number;
  message: string;
  error?: string;
  async: boolean;
  processedModelUrl?: string;
  unfoldSvgUrl?: string;
  unfoldPdfUrl?: string;
  unfoldZipUrl?: string;
  resultStatus?: ProjectStatus;
  stats?: ProcessStats;
  craftability?: CraftabilityScore;
  exportBlocked?: boolean;
  hasUnfoldOverlap?: boolean;
};

export type PollOptions = {
  intervalMs?: number;
  timeoutMs?: number;
  signal?: AbortSignal;
  onProgress?: (job: ProcessJobResponse) => void;
  /** Force HTTP polling (skip SSE). */
  preferPoll?: boolean;
};

const DEFAULT_INTERVAL_MS = 1500;
const DEFAULT_TIMEOUT_MS = 900_000;

function processJobFailure(job: ProcessJobResponse): Error | null {
  if (job.status === "cancelled") {
    return new ProcessJobCancelledError(job.error ?? "Processing cancelled.");
  }
  if (job.status === "failed") {
    return new Error(job.error ?? "Papercraft processing failed.");
  }
  return null;
}

function streamProcessJob(
  jobId: string,
  options: PollOptions = {},
): Promise<ProcessJobResponse> {
  const { timeoutMs = DEFAULT_TIMEOUT_MS, signal, onProgress } = options;

  return streamJobEvents<ProcessJobResponse>(
    `/api/process-jobs/${jobId}/events`,
    {
      timeoutMs,
      signal,
      onMessage: onProgress,
      isTerminal: (job) => job.status === "completed",
      isFailure: processJobFailure,
    },
  );
}

async function pollProcessJobHttp(
  jobId: string,
  options: PollOptions = {},
): Promise<ProcessJobResponse> {
  const intervalMs = options.intervalMs ?? DEFAULT_INTERVAL_MS;
  const timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  const { signal } = options;
  const started = Date.now();

  while (Date.now() - started < timeoutMs) {
    throwIfAborted(signal);

    const response = await fetch(`/api/process-jobs/${jobId}`, {
      signal,
      headers: apiAuthHeaders(),
    });
    if (!response.ok) {
      const body = (await response.json().catch(() => ({}))) as { detail?: string };
      throw new Error(body.detail ?? "Failed to fetch process job status.");
    }

    const job = (await response.json()) as ProcessJobResponse;
    throwIfAborted(signal);
    options.onProgress?.(job);

    const failure = processJobFailure(job);
    if (failure) throw failure;
    if (job.status === "completed") return job;

    await abortableDelay(intervalMs, signal);
  }

  throw new Error("Processing timed out. Please try again later.");
}

export async function pollProcessJob(
  jobId: string,
  options: PollOptions = {},
): Promise<ProcessJobResponse> {
  if (options.preferPoll) {
    return pollProcessJobHttp(jobId, options);
  }

  return withStreamOrFallback(
    () => streamProcessJob(jobId, options),
    () => pollProcessJobHttp(jobId, options),
  );
}

File: apps/api/app/schemas/process_job.py
"""Async papercraft processing job schemas."""

from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.job import JobStatus
from app.schemas.model import ProjectSettings, ProjectStatus
from app.schemas.stats import CraftabilityScore, ProcessStats


class ProcessJob(BaseModel):
    """Tracks async papercraft pipeline work for a project."""

    id: str
    project_id: str = Field(alias="projectId")
    status: JobStatus = JobStatus.QUEUED
    progress: int = 0
    message: str = "Queued"
    error: str | None = None
    settings: ProjectSettings
    project_name: str = Field(alias="projectName")
    source_path: str = Field(alias="sourcePath")
    processed_model_url: str | None = Field(alias="processedModelUrl", default=None)
    unfold_svg_url: str | None = Field(alias="unfoldSvgUrl", default=None)
    unfold_pdf_url: str | None = Field(alias="unfoldPdfUrl", default=None)
    unfold_zip_url: str | None = Field(alias="unfoldZipUrl", default=None)
    result_status: ProjectStatus | None = Field(alias="resultStatus", default=None)
    stats: ProcessStats | None = None
    craftability: CraftabilityScore | None = None
    export_blocked: bool = Field(alias="exportBlocked", default=False)
    has_unfold_overlap: bool = Field(alias="hasUnfoldOverlap", default=False)
    cancel_requested: bool = Field(alias="cancelRequested", default=False)
    locked_by: str | None = Field(alias="lockedBy", default=None)
    locked_until: datetime | None = Field(alias="lockedUntil", default=None)
    attempts: int = 0
    last_error: str | None = Field(alias="lastError", default=None)
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")

    model_config = {"populate_by_name": True}


class ProcessJobResponse(BaseModel):
    job_id: str = Field(alias="jobId")
    project_id: str = Field(alias="projectId")
    status: JobStatus
    progress: int = 0
    message: str = "Queued"
    error: str | None = None
    async_mode: bool = Field(alias="async", default=True)
    processed_model_url: str | None = Field(alias="processedModelUrl", default=None)
    unfold_svg_url: str | None = Field(alias="unfoldSvgUrl", default=None)
    unfold_pdf_url: str | None = Field(alias="unfoldPdfUrl", default=None)
    unfold_zip_url: str | None = Field(alias="unfoldZipUrl", default=None)
    result_status: ProjectStatus | None = Field(alias="resultStatus", default=None)
    stats: ProcessStats | None = None
    craftability: CraftabilityScore | None = None
    export_blocked: bool = Field(alias="exportBlocked", default=False)
    has_unfold_overlap: bool = Field(alias="hasUnfoldOverlap", default=False)
    cancel_requested: bool = Field(alias="cancelRequested", default=False)

    model_config = {"populate_by_name": True}
